src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,241 @@
+            putWithCollision(newNode);
+        }
+        if (size == threshold) {
+            resize();
+        }
     }
 
     @Override
     public V getValue(K key) {
+        V value;
+        if (key == null) {
+            return getValueByNullKey(key);
+        }
+        for (Node<K, V> nodeForGetValue : table) {
+            if (nodeForGetValue != null && key.equals(getKey(nodeForGetValue))) {
+                value = nodeForGetValue.value;
+                return value;
+            }
+            if (nodeForGetValue != null && nodeForGetValue.next != null) {
+                Node<K, V> nextNode = nodeForGetValue.next;
+                while (nextNode != null) {
+                    if (getKey(nextNode) == key
+                            || (getKey(nextNode) != null && getKey(nextNode).equals(key))) {
+                        value = nextNode.value;
+                        return value;
+                    }
+                    nextNode = nextNode.next;
+                }
+            }
+        }
         return null;
     }
 
+    @Override
+    public V[] values() {
+        V[] valuesArray = (V[]) new Object[size];
+        int index = 0;
+        for (Node<K, V> node : table) {
+            if (node != null) {
+                valuesArray[index] = node.value;
+                index++;
+            }
+            if (node != null && node.next != null) {
+                Node<K, V> nextNode = node.next;
+                while (nextNode != null) {
+                    valuesArray[index] = nextNode.value;
+                    index++;
+                    nextNode = nextNode.next;
+                }
+            }
+            if (index >= size) {
+                break;
+            }
+        }
+        return valuesArray;
+    }
+
+    @Override
+    public K[] keySet() {
+        K[] keysArray = (K[]) new Object[size];
+        int index = 0;
+        for (Node<K, V> node : table) {
+            if (node != null) {
+                keysArray[index] = getKey(node);
+                index++;
+            }
+            if (node != null && node.next != null) {
+                Node<K, V> nextNode = node.next;
+                while (nextNode != null) {
+                    keysArray[index] = nextNode.key;
+                    index++;
+                    nextNode = nextNode.next;
+                }
+            }
+            if (index >= size) {
+                break;
+            }
+        }
+        return keysArray;
+    }
+
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    @Override
+    public void clear() {
+        final int sizeAfterClear = 0;
+        if (!isEmpty()) {
+            for (int i = 0; i < table.length; i++) {
+                table[i] = null;
+            }
+            size = sizeAfterClear;
+        }
+    }
+
+    @Override
+    public boolean containsKey(K key) {
+        Node<K, V> currentNode;
+        for (Node<K, V> node : table) {
```

Comment:
I think it's better to calculate in what bucket we should look for this key (not go through all nodes)